receiver/nsxtreceiver/scraper.go

```diff
@@ -0,0 +1,271 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nsxtreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/nsxtreceiver"
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/receiver/scrapererror"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/nsxtreceiver/internal/metadata"
+	dm "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/nsxtreceiver/internal/model"
+)
+
+type scraper struct {
+	config   *Config
+	settings component.TelemetrySettings
+	host     component.Host
+	client   Client
+	mb       *metadata.MetricsBuilder
+	logger   *zap.Logger
+}
+
+func newScraper(cfg *Config, settings component.TelemetrySettings) *scraper {
+	return &scraper{
+		config:   cfg,
+		settings: settings,
+		mb:       metadata.NewMetricsBuilder(cfg.Metrics),
+		logger:   settings.Logger,
+	}
+}
+
+func (s *scraper) start(ctx context.Context, host component.Host) error {
+	s.host = host
+	err := s.ensureClient()
+	// allow reconnectivity
+	if err != nil {
+		return fmt.Errorf("unable to ensure client connectivity: %s", err)
+	}
+	return nil
+}
+
+type nodeClass int
+
+const (
+	transportClass nodeClass = iota
+	managerClass
+)
+
+func (s *scraper) scrape(ctx context.Context) (pdata.Metrics, error) {
+	errs := &scrapererror.ScrapeErrors{}
+	if err := s.ensureClient(); err != nil {
+		errs.Add(err)
+		return pmetric.NewMetrics(), errs.Combine()
+	}
+
+	r := s.retrieve(ctx, errs)
```

Comment:
Since this is the only place where errors are generated, this func should instantiate and return `errs`.